docs/proposals/ldap-group-sync.md

```diff
@@ -0,0 +1,92 @@
+
+## Supporting Group Sources
+In order to link `Group` structures to external providers of their data, each `Group` will need an (optional) `GroupSource`:
+
+```go
+type Group struct{
+  ...
+  // Source is an optional parameter that enables the Group member list to be linked to an entry in another system or database
+  Source GroupSource
+}
+
+// GroupSource enables the Group member list to be linked to an entry in another system or database
+type GroupSource interface{
+  parent *Group
+  // Sync updates the group's list of members to reflect the parallel list in the remote system or database
+  Sync() (deleted bool, members []string, err error)
+}
+```
+
+## LDAP Sync
+
+### Schemas To Support
+The LDAP source may look like this, where the `LDAPMemberSource` may be backed by different implementations in order to allow for differing group schemas in the customer LDAP server:
+
+```go
+type LDAPGroupSource interface{
+    parent *Group
+    memberSource LDAPMemberSource
+}
+
+func (l *LDAPGroupSource)Sync() (deleted bool, members []string, err error){
+  // use l.memberSource.FetchMembers() to get LDAP version of member list
+  // use l.parent.AddMember(), RemoveMember() to update parent member list
+}
+
+// LDAPMemberSource is a strategy for getting a list of users representing the membership of a group from an LDAP server
+type LDAPMemberSource interface {
+  FetchMembers() (members []string, err error)
+}
+```
+
+For instance, an LDAP schema where groups are first-class members may lead to a `LDAPMemberSource` like this:
+
+```go
+// LDAPGroupMemberSource gets the members of a group from an LDAP server where groups are first-class objects and their members are encoded using an attribute.
+type LDAPGroupMemberSource struct {
+  // filter is an opaque string holding the LDAP query that returns the group this source
+  filter string
+  // memberListAttribute is an opaque string holding the attribute of the group returned by the LDAP server that holds the list of members
+  memberListAttribute string
+}
+
+func (l *LDAPGroupMemberSource)FetchMembers() (members []string, err error) {
+  // use l.filter to get first-class group
+  // use l.memberListAttribute to extract members from the group
+}
+```
+
+LDAP schemas that encode groups not as first-class objects but as attributes on members will lead to an `LDAPMemberSource` like this:
+
+```go
+// LDAPUserListMemberSource gets the members of a group from an LDAP server where groups are encoded using an attribute on users.
+type LDAPUserListMemberSource struct {
+  // filter is an opaque string holding the LDAP query that returns the members for the group holding this source
+  filter string
+}
+
+func (l *LDAPUserListMemberSource)FetchMembers() (members []string, err error) {
+  // use l.filter to get list of users
+}
+```
+
+### Tuneable Parameters
```

Comment:
You need to allow admins to specify additional filters about the groups they want to sync (they might not want the entire tree).